x/mongo/driver/topology/server.go

```diff
@@ -437,119 +526,157 @@ func (s *Server) updateDescription(desc description.Server) {
-				// that wasn't successfully created
-				_ = conn.close()
-			}
+	s.conn.connect(s.heartbeatCtx)
+	return s.conn.wait()
+}
+
+// cancelCheck cancels in-progress connection dials and reads. It does not set any fields on the server.
+func (s *Server) cancelCheck() {
+	var conn *connection
+
+	// Take heartbeatLock for mutual exclusion with the checks in the update function.
+	s.heartbeatLock.Lock()
+	if s.heartbeatCtx != nil {
+		s.heartbeatCtxCancel()
+	}
+	conn = s.conn
+	s.heartbeatLock.Unlock()
+
+	if conn == nil {
+		return
+	}
+
+	// If the connection exists, we need to wait for it to be connected. We can ignore the error from conn.wait(). If
+	// the connection wasn't successfully opened, its state was set back to disconnected, so calling conn.close() will
+	// be a noop.
+	conn.closeConnectContext()
+	_ = conn.wait()
+	_ = conn.close()
+}
+
+func (s *Server) checkWasCancelled() bool {
+	return s.heartbeatCtx.Err() != nil
+}
+
+func (s *Server) createBaseOperation(conn driver.Connection) *operation.IsMaster {
+	return operation.
+		NewIsMaster().
+		ClusterClock(s.cfg.clock).
+		Deployment(driver.SingleConnectionDeployment{conn})
+}
+
+func (s *Server) check() (description.Server, error) {
+	var descPtr *description.Server
+	var err error
+
+	// Create a new connection if this is the first check, the connection was closed after an error during the previous
+	// check, or the previous check was cancelled.
+	if s.conn == nil || s.conn.closed() || s.checkWasCancelled() {
+		// Create a new connection and add it's handshake RTT as a sample.
+		err = s.setupHeartbeatConnection()
+		if err == nil {
+			// Use the description from the connection handshake as the value for this check.
+			s.rttMonitor.addSample(s.conn.isMasterRTT)
+			descPtr = &s.conn.desc
 		}
+	}
+
+	if descPtr == nil && err == nil {
+		// An existing connection is being used. Use the server description properties to execute the right heartbeat.
+
+		// Wrap conn in a type that implements driver.StreamerConnection.
+		heartbeatConn := initConnection{s.conn}
+		baseOperation := s.createBaseOperation(heartbeatConn)
+		previousDescription := s.Description()
+
+		switch {
+		case s.conn.getCurrentlyStreaming():
+			// The connection is already in a streaming state, so we stream the next response.
+			err = baseOperation.StreamResponse(s.heartbeatCtx, heartbeatConn)
+		case previousDescription.TopologyVersion != nil:
+			// The server supports the streamable protocol. Set the socket timeout to
+			// connectTimeoutMS+heartbeatFrequencyMS and execute an awaitable isMaster request. Set conn.canStream so
+			// the wire message will advertise streaming support to the server.
+
+			// Calculation for maxAwaitTimeMS is taken from time.Duration.Milliseconds (added in Go 1.13).
+			maxAwaitTimeMS := int64(s.cfg.heartbeatInterval) / 1e6
+			s.conn.setSocketTimeout(s.cfg.heartbeatTimeout + s.cfg.heartbeatInterval)
```

Comment:
Does this handle the `connectTimeoutMS=0` case added in [SPEC-1705](https://jira.mongodb.org/browse/SPEC-1705)?

Reply:
SGTM. ([GODRIVER-1656](https://jira.mongodb.org/browse/GODRIVER-1656) exists now.)

Reply:
Added the changes for GODRIVER-1656 in a new commit. I verified that the tests fail without the changes and pass with them.